### ISSUE TITLE
Security updates are not applied in the Dockerfile

### ISSUE THREAD

**Issue by kkoppel** [nextcloud/docker#1247]
Currently the Docker images for Nextcloud do not apply available upstream security updates when the image is being built. E.g. currently the latest stable-apache image, which was built just 2 hours ago, does not have this Debian security fix (released on August 31st) installed:
https://www.debian.org/security/2020/dsa-4757

IMO, it would be much easier and safer for everybody to apply any available upstream package updates while building the official image, rather than forcing users to apply updates inside the image or build their own image based on the official one, just to be sure that security updates are installed.

I fully understand and have no problem with the explanation given for #623, but I do not think that it is a good idea to rely on upstream Docker images having these updates applied promptly.

For Debian-based images this should be sufficient:
```
diff --git a/19.0/apache/Dockerfile b/19.0/apache/Dockerfile
index c77835a..034e6a8 100644
--- a/19.0/apache/Dockerfile
+++ b/19.0/apache/Dockerfile
@@ -5,6 +5,7 @@ FROM php:7.4-apache-buster
 RUN set -ex; \
     \
     apt-get update; \
+    apt-get upgrade -y ; \
     apt-get install -y --no-install-recommends \
         rsync \
         bzip2 \
```

For Alpine-based images I believe the respective change would be:
```
diff --git a/19.0/fpm-alpine/Dockerfile b/19.0/fpm-alpine/Dockerfile
index 55a0b20..ab23617 100644
--- a/19.0/fpm-alpine/Dockerfile
+++ b/19.0/fpm-alpine/Dockerfile
@@ -4,6 +4,8 @@ FROM php:7.4-fpm-alpine3.12
 # entrypoint.sh and cron.sh dependencies
 RUN set -ex; \
     \
+    apk update --no-cache; \
+    apk upgrade --no-cache; \
     apk add --no-cache \
         rsync \
     ; \
```

**Comment by Derkades** [comment]
Doing `apt upgrade` or the equivalent for other package managers is considered bad practice. https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#run

**Comment by kkoppel** [comment]
As I understand that doc, they are advising against it due to the possibility of some Debian "essential" package upgrades failing when run inside a Docker container. Obviously the base image should be updated instead, but IRL that is not always possible.

Doing the "apt-get upgrade" will not break anything in the vast majority of cases, because either there are no updates at all or there is an update that will apply just fine. If it does break, then this is a good way to find out that the base image is not up to date. In fact I would probably prefer not building a new image at all until the base image is updated with that essential package security update, because this brings attention to the problem.

IMO, having security updates applied is far more important than avoiding the potential extra hassle of occasionally dealing with breakage.

**Comment by JustACasual** [comment]
Any update on this issue? I was also wondering why I get a 'not patched' warning from https://scan.nextcloud.com after using the :latest nextcloud docker image...

**Comment by codingkrabbe** [comment]
I also face the issue right now that I basically cannot use the docker image, as it is outdated :-( 
This is a security issue for any deployed instance that is based on the docker image, please fix this.

**Comment by StephenLittBelch** [comment]
I agree with the sentiment, although I am not sure if there is need for concern in this case. The vulnerabilities in https://www.debian.org/security/2020/dsa-4757 mainly affect configurations using HTTP/2. This usually involves using the _*-fpm_ Nextcloud images, which would mean using a separate (up to date) Apache/httpd image.

Out of date images are somewhat inherent with the stacked-build design of Docker images, when you are not building all of the layers yourself. It is possible to run `docker exec [container] apt update && apt upgrade -y` within a container if ever needed. More cleanly, building a custom Dockerfile that inherits the out of date _*-apache_ Nextcloud image and just runs `apt upgrade -y` would solve the issue until the official image catches up.

**Comment by codingkrabbe** [comment]
Hi Stephen,
Thanks for the answer, however I cannot agree with you here. 
"Out of date images are somewhat inherent with the stacked-build design of Docker images, when you are not building all of the layers yourself." - I don't see this like you, you could always add the upgrade step in the dockerfile as a separate layer, which makes sure that everything is up to date.

"and just runs apt upgrade -y for use until the official image catches up." - but this goes against everything that you want to create with docker, namely an infrastructure without configuration drift and the ability to recreate the infra from scratch and having everything stored properly as Config as Code. If I have to start to patch my docker containers, the whole setup is basically doomed from the start.

While this vulnerability might not affect this image, you cannot make sure that this is also true for all other vulnerabilities. 
Also nobody expects that he has to manually patch their containers, this goes against best practices.
See also here:
- https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#create-ephemeral-containers
- https://blogs.tunelko.com/wp-content/uploads/2017/08/CIS_Docker_1.13.0_Benchmark_v1.0.0.pdf --> 4.4 specifically mentions a rebuild,
- https://snyk.io/blog/10-docker-image-security-best-practices/ --> fix as part of the Docker CI
- https://serverfault.com/questions/924394/is-it-a-good-practice-to-put-update-upgrade-statements-inside-the-dockerfile --> use up-to-date base image or update yourself if the base image part is not possible

I really like nextcloud and would like to use it in the public internet, it would be really great if you could enhance the security process here and provide up to date base images :-)

**Comment by StephenLittBelch** [comment]
> "and just runs apt upgrade -y for use until the official image catches up." - but this goes against everything that you want to create with docker, namely an infrastructure without configuration drift and the ability to recreate the infra from scratch and having everything stored properly as Config as Code. If I have to start to patch my docker containers, the whole setup is basically doomed from the start.

I completely agree. Just reflecting on my personal experience using the images and manually updating them for critical vulnerabilities with OpenSSL, Apache, PHP or Nextcloud a few times before a new Nextcloud version or Dockerfile change triggered an updated image. The Nextcloud images do look to rebuild the _php:7.4-apache-buster_ images they are based upon: [Image Layer Details - nextcloud:19.0.3-apache](https://hub.docker.com/layers/nextcloud/library/nextcloud/19.0.3-apache/images/sha256-a1016e24c9acedc68977bff797794e9b0273e632fb942b44adb0766ec5d60355).

**Comment by JustACasual** [comment]
Can't quite follow your discussions, guys. Can you summarize for me? I'm just a casual :D 
Will the docker run nextcloud:latest result in a fully patched nextcloud or not in the future?

**Comment by kkoppel** [comment]
@JustACasual This issue is about ensuring that when the Nextcloud Docker image is built, then it has all the currently available security updates applied. What happens an hour/day/week/year after the image is built is a whole other discussion.

**Comment by J0WI** [comment]
In our layer, we're just maintaining security updates for Nextcloud itself. All other dependencies, namely Apache or PHP, are maintained here: https://github.com/docker-library/php
As soon as there is a new version for any of the base images all Nextcloud images are rebuild automatically.

**Comment by kkoppel** [comment]
@J0WI that's nice and I can see how in theory this should give up to date images, but as we have seen in reality that's not always the case.
I don't really understand what would be the downside of ensuring that all available security updates are installed at image build time?

**Comment by J0WI** [comment]
> I don't really understand what would be the downside of ensuring that all available security updates are installed at image build time?

Docker images use filesystem layers. Running `apk|apt upgrade` destroys all previous layers.

> that's nice and I can see how in theory this should give up to date images, but as we have seen in reality that's not always the case.

That's a general issue with images, packages an last but not least distros. We do our best to keep at least our layer up to date.

**Comment by codingkrabbe** [comment]
so some example from our setup:
- we pull every night the latest container from nextcloud and apply it to our infrastructure
- right now, the open-ssl lib is outdated, the latest version to have is already for 5 days available
- image we use: nextcloud:latest

root@860a84757e6b:/var/www/html/themes# apt list --upgradeable
Listing... Done
libssl1.1/stable 1.1.1d-0+deb10u5 amd64 [upgradable from: 1.1.1d-0+deb10u4]
libzstd1/stable 1.3.8+dfsg-3+deb10u2 amd64 [upgradable from: 1.3.8+dfsg-3]
openssl/stable 1.1.1d-0+deb10u5 amd64 [upgradable from: 1.1.1d-0+deb10u4]

So if I see this right, we have the latest container and still not all relevant security patches applied correct?
Which can be critical for an environment in the internet.
Also, although we nightly pull the latest image and apply it, we get a security warning from scan.nextcloud.com:

![image](https://user-images.githubusercontent.com/34913588/108589666-b9ba1700-735f-11eb-8c1a-4ab2dd65b153.png)

Is there something wrong in the setup? Or are nextcloud patches also applied later?

**Comment by codingkrabbe** [comment]
so, the issue is now 2 years old, I currently gave up on the topic :D

What I remember from back then and what we accepted as truth is basically, that the container version always lags a bit behind the non containerized version, when pulling the image directly. 

This was always just true for a few days, and didn't happen for security critical patches as far as I know, so it's fine for my (low-sec) use case.

I am also not able to find any documentation on the synchronization between the different releases (traditional vs. the fpm image vs. the base image) is there any policy?
